Array ufunc fixups

mhvk opened the pull request.

eric-wieser commented on numpy/core/src/umath/override.c:
```diff
@@ -180,18 +179,6 @@ normalize_reduceat_args(PyUFuncObject *ufunc, PyObject *args,
 static int
-normalize_outer_args(PyUFuncObject *ufunc, PyObject *args,
-                    PyObject **normal_args, PyObject **normal_kwds)
-{
```
I might have been tempted to keep this as a simple wrapper that forwards to `__call__`, with a brief comment explaining they are the same. The compiler should inline the single function call anyway, right?

mhvk replied:
But I already combine `reduce` and `accumulate` similarly. But let me at least adjust the function name accordingly.

eric-wieser replied:
You'll need to split those up again when numpy/numpy#8187 is merged though...

mhvk replied:
Yikes, reduce and accumulate are really not quite the same, despite my comment. What was I thinking...

eric-wieser replied:
Clearly the same thing as whoever decided that adding the `keepdims` argument in the first place was a good idea ;)

mhvk replied:
Actually, never mind, it is OK. Sigh. Just the comment is wrong.

mhvk replied:
Never mind the never mind...  in `ufunc_object`, `accumulate` happily accepts `keepdims`. This seems like a mistake.

Anyway, I'm clearly too tired for this, and better stop now.

eric-wieser replied:
> This seems like a mistake.

This is a mistake that we now have a deprecation warning for, and is scheduled for removal in 1.14

charris replied:
Do we want to separate the handling of reduce and accumulate here with the latter dropping keepdims?

mhvk commented:
@eric-wieser, @charris - OK, I split `reduce` and `accumulate`, and also split `outer` and `__call__` again, since the current implementation for the standard path for `outer` explicitly insists that the positional arguments only include input arguments (unlike for `__call__`), and I thought that might as well be kept (otherwise things that work with an override would not work with the standard path). Indeed, as it was, the call to `CheckOverride` was after checking that there were exactly two arguments, so the original override was correct. But that also seemed silly: for all the other ufunc methods, `override` is done as soon as possible, and I didn't see why that shouldn't be the case for `outer`; if someone wants to override with 3 inputs, that should be allowed, I think.

Anyway, I ended up adding number of argument checks for all the other methods...  I think it is now OK. It may well be possible to optimize a bit better, but my feeling is that it is not worth doing that right now, since, really, the normalization should be combined between the normal and override paths.

eric-wieser commented on numpy/core/src/umath/override.c:
```diff
@@ -96,23 +108,72 @@ normalize___call___args(PyUFuncObject *ufunc, PyObject *args,
+                Py_DECREF(obj);
+            }
+        }
+    }
```
The chunk of code here that performs arg-to-kwarg mappings is repeated rather a lot, isn't it? Can a helper function of some kind be made?

mhvk replied:
Well, that was why I had `reduce` and `accumulate` together... I agree that it could be done better with helper functions, but would suggest to leave this for a follow-up PR -- which then could also address the duplication between override and normal paths (see #8892). I wouldn't be surprised is that made whatever I do here rather irrelevant...

eric-wieser replied:
Yeah, follow up PR seems fine

eric-wieser commented:
> really, the normalization should be combined between the normal and override paths.

Yeah, I agree - it'd probably be better to just rebuild the argument tuple from the fully parsed values, rather than try and mangle the inputs into the right things

mhvk commented:
Indeed, I had been wondering about that too... Especially given that a new `kwargs` dict is constructed here anyway, it is not at all obvious that it would be much less efficient.

charris commented:
Merged, thanks Marten.
